src/Aspire.Dashboard/Components/Controls/LogViewer.razor.cs

```diff
@@ -46,31 +58,76 @@ await foreach (var batch in batches.WithCancellation(cancellationToken))
                 continue;
             }
 
-            List<LogEntry> entries = new(batch.Count);
-
             foreach (var (content, isErrorOutput) in batch)
             {
-                entries.Add(logParser.CreateLogEntry(content, isErrorOutput));
+                InsertSorted(_logEntries, logParser.CreateLogEntry(content, isErrorOutput));
             }
 
-            await _jsModule!.InvokeVoidAsync("addLogEntries", cancellationToken, entries);
+            StateHasChanged();
         }
     }
 
-    internal async Task ClearLogsAsync(CancellationToken cancellationToken = default)
+    private static void InsertSorted(List<LogEntry> logEntries, LogEntry logEntry)
     {
-        await _cancellationSeries.ClearAsync();
+        if (logEntry.ParentId != null)
+        {
+            // If we have a parent id, then we know we're on a non-timestamped line that is part
```

Comment:
Was hoping the need to do this interleaving would go away at some point. I guess that some point is not yet...

Reply:
I just replicated what JS was doing. I'm not a fan of it. Reordering data from the server just before it is rendered is probably the wrong place. If reordering is required, it should happen before storing the data inside the host.

Can you talk more about why it is required? What was the original issue?

Reply:
Yeah, sorry, I probably could have been more clear on that comment. I had hoped it was going to go away with streaming logs, but @karolz-ms indicated there were still two separate streams so there might be an issue.

The reordering was required because the logs arrived out of order from the two different streams (stdout and stderr).  This was especially true with logs being written to the file system - the first stderr line could have been seconds or minutes after the first stdout line, but since the two files were read at the same time, we'd get them in unpredictable order. Though whatever the source is (docker logs acted slightly differently) they'd still need to be interleaved at some point.

Looking through the current code on the dashboard side, I don't see anywhere obvious where we see the two different streams anymore. Maybe @karolz-ms can speak to how that works in the streaming world now?

Reply:
There are still two streams, they now come from k8s service:

https://github.com/dotnet/aspire/blob/0783a609d369634c1a0aa5d399b78c79a52caafe/src/Aspire.Hosting/Dcp/ResourceLogSource.cs#L28-L29

I think this would still be a problem.

Reply:
DCP could combine the streams (and apply timestamps to Executable logs), but then we need to figure out how to convey what entries come from stdout vs stderr. E.g. change the format to JSON lines or something similar, but then there is a question of how to deal with non-printable characters etc.

Reply:
If DCP combined the streams, would they be in the correct order? How does DCP get its data?

Reply:
> E.g. change the format to JSON lines or something similar, but then there is a question of how to deal with non-printable characters etc.

I guess a JSON serializer would escape them to /u342345 format.

Reply:
For Containers the logs arrive in the order the container orchestrator is providing them. But since we are using the CLI to talk to the orchestrator, it boils down to reading `stdout` and `stderr`. 

The same applies to Executables. When DCP launches Executables directly, it is also responsible for reading `stdout` and `stderr` streams. When using the IDE to launch Executables (i.e. dealing with VS projects), it is relying on the IDE to provide the logs. 

Bottom line this is probably sufficient for UI to have logs emitted ROUGHLY at the same time to surface close in the combined stream. But there will definitively be some jitter involved, and if the expectation is that developers will be able to infer the order in which things happened based on the order in which they appeared in the (combined) log stream, especially if we are talking about microseconds delay between writes to `stdout` vs `stderr`, that is probably a stretch.